### PR TITLE
Add ftse  selection regret example notebook

### PR DESCRIPTION
**Pull request by Shr10122**
## What changes

Adds examples/selection_regret_ftse.ipynb, a UK enquities(FTSE 100)
example following the existing selection-regret notebook pattern.
It Compares KFold(shuffle=False), KFold(shuffle=True),PurgedGroupKFold, and WalkForwardGroupKFold with a held-out 2024 test period.

## Checklist

- [x] `ruff check . && black --check src tests tools examples/_lcl_harness.py` is clean.
- [x] `mypy src tests` is clean.
- [x] `pytest -q` is green locally.

## Notes for the reviewer

It's a first working version, opened for early feedback pr, your suggestion on the issue. A few things i would like to input on:

- The result is a null result(-0.5 AUC across all CV methods, including the untouched 2024 holdout). Leakage did not visibly inflate results here ,likely because the 3 features used (5d/20d returns, 20d volatility) are too weak for leaky cv to have  much to latch onto.
- not yet checked against the prose gate/docs build will confirm before marking this ready for review.
-Open to feedback on whether this result is sufficient as  is or whether I should add a feature more likely to show leaky vs safe divergence before finanlizing.

### CODE REVIEW

File: examples/selection_regret_ftse.ipynb [hunk 1]
```diff
@@ -0,0 +1,721 @@
...
+    "# purgedcv needs: prediction_times (when we make prediction), evaluation_times (when label is known)\n",
+    "prediction_times = pd.Series(model_selection_data['Date'].values, index=range(len(model_selection_data)))\n",
+    "evaluation_times = pd.Series(model_selection_data['evaluation_times'].values, index=range(len(model_selection_data)))\n",
+    "\n",
+    "# Also need to group by date for PurgedGroupKFold (to keep same-day rows together)\n",
+    "groups = pd.Series(model_selection_data['Date'].values, index=range(len(model_selection_data)))\n",
+    "\n",
+    "# Standardize features\n",
+    "scaler = StandardScaler()\n",
+    "X_scaled = scaler.fit_transform(X)\n",
```
Comment (eslazarev) [inline examples/selection_regret_ftse.ipynb:408]: This fits the scaler on the validation rows too, including for the methods marked SAFE. Please put it in a Pipeline with LogisticRegression and fit the pipeline separately in each fold.

File: examples/selection_regret_ftse.ipynb [hunk 2]
```diff
@@ -0,0 +1,721 @@
...
+    "    'method': 'KFold(shuffle=True)',\n",
+    "    'mean_auc': np.mean(aucs_shuffled),\n",
+    "    'std_auc': np.std(aucs_shuffled),\n",
+    "})\n",
+    "\n",
+    "# 3. PurgedGroupKFold — purges overlapping labels, groups by date\n",
+    "print(\"\\n3. PurgedGroupKFold — purged + grouped by date (SAFE)\")\n",
+    "# The data is currently ordered by ticker, so prediction_times resets when\n",
+    "# moving from one ticker to the next. Reorder every aligned object by time.\n",
+    "order = np.argsort(prediction_times.to_numpy(), kind=\"stable\")\n",
```
Comment (eslazarev) [inline examples/selection_regret_ftse.ipynb:458]: Could you move this sorting step before all the CV comparisons? Right now, unshuffled KFold gets ticker-ordered data, so its folds mostly separate stocks rather than time periods.

An unpurged baseline with the same test-date blocks as PurgedGroupKFold would also help us see what purging actually changes.

File: examples/selection_regret_ftse.ipynb [hunk 3]
```diff
@@ -0,0 +1,721 @@
...
+    "\n",
+    "# For each unique date in our data, find the date 5 business days later\n",
+    "eval_times_dict = {}\n",
+    "for date in sorted(data['Date'].unique()):\n",
+    "    try:\n",
+    "        # Find the position of this date in the business-day index\n",
+    "        pos = bday_index.get_loc(date)\n",
+    "        # 5 business days ahead = current position + 5\n",
+    "        if pos + 5 < len(bday_index):\n",
+    "            eval_times_dict[date] = bday_index[pos + 5]\n",
```
Comment (eslazarev) [inline examples/selection_regret_ftse.ipynb:228]: Five weekdays isn't always five observations for a stock. For example, starting on March 28, 2024, this gives April 4, but the fifth next observation is April 8.

Please use a per-ticker shift(-5) for both the future close and evaluation_times, before dropping incomplete rows.

File: examples/selection_regret_ftse.ipynb [hunk 4]
```diff
@@ -0,0 +1,721 @@
...
+    "\n",
+    "cv_wf = WalkForwardSplit(\n",
+    "    n_splits=5,\n",
+    "    test_size=test_size,\n",
+    "    prediction_times=prediction_times,\n",
+    "    evaluation_times=evaluation_times,\n",
+    ")\n",
+    "aucs_wf = []\n",
+    "# WalkForwardSplit requires n_splits * test_size < n_samples.\n",
+    "test_size = (len(X_scaled) - 1) // 5\n",
```
Comment (eslazarev) [inline examples/selection_regret_ftse.ipynb:504]: With 43,410 rows, this leaves just five training rows before the first test fold. Purging removes them, so the first fold gets skipped.

Let's reserve a reasonable initial training period and put the test boundaries between complete dates. Then check that all five folds have usable training data.

File: examples/selection_regret_ftse.ipynb [hunk 5]
```diff
@@ -0,0 +1,721 @@
...
+      "  - Overfitting during model selection, or\n",
+      "  - A genuine regime change in 2024, or\n",
+      "  - Small-sample noise (holdout has fewer rows than full CV)\n"
+     ]
+    }
+   ],
+   "source": [
+    "# we already did model comparison above, this is the final check)\n",
+    "final_model = LogisticRegression(max_iter=1000)\n",
+    "final_model.fit(X_scaled, y)\n",
```
Comment (eslazarev) [inline examples/selection_regret_ftse.ipynb:603]: We're comparing CV scores here, but we're not actually selecting a model: every method uses the same LogisticRegression.

For a selection-regret example, please define a small candidate set, let each CV method choose its winner, then refit those winners and evaluate them on the same holdout. It's fine if they all choose the same model.

Alternatively, we could keep the simpler experiment and rename it as a comparison of CV estimates.

**Review by eslazarev — COMMENTED** [review]
Thanks for putting this together. I reproduced your reported scores in a separate script, so I don't think the numbers themselves are the issue. A result close to 0.5 is fine.

There are a few things in the setup I'd like to fix before merging. I've left comments below.

I also tried a small model-selection experiment. Shuffled KFold picked a boosting model that did worse on the holdout, while the other methods picked logistic regression. But ordinary date blocks and purged CV picked the same model, and the holdout difference was uncertain. So there's no need to make this example show a clear win for purging.

File: examples/selection_regret_ftse.ipynb (hunk 1 above)
Comment (Shr10122) [reply examples/selection_regret_ftse.ipynb]: Fixed moved scaler+model into single Pipeline(make pipeline()), fit fresh inside each fold. Scaler now only sees that fold's training rows.

File: examples/selection_regret_ftse.ipynb (hunk 2 above)
Comment (Shr10122) [reply examples/selection_regret_ftse.ipynb]: Now sorting by Date before any CV comparisons run. Also added the same date blocks withoit purge baseline you suggested, it comes in close to PuredGroupKFold(0.5158 vs 0.516), well within noise.

File: examples/selection_regret_ftse.ipynb (hunk 3 above)
Comment (Shr10122) [reply examples/selection_regret_ftse.ipynb]: Replaced the bday_index/get_loc approach entirely with a per-ticker .shift(-5) on the actual price series this respects the real trading  calendar automatically since it only ever uses rows that actually traded, no calendar arithmetic needed.

File: examples/selection_regret_ftse.ipynb (hunk 4 above)
Comment (Shr10122) [reply examples/selection_regret_ftse.ipynb]: Now reserving an initial ~1 year training only period before computing test_size from the remaining rows, so early folds have real training data left after purging.

File: examples/selection_regret_ftse.ipynb (hunk 5 above)
Comment (Shr10122) [reply examples/selection_regret_ftse.ipynb]: Rebuilt this section as a real selection experiment across 3 candidates(logisticRegression, Randomforest shallow/deep). Result: shuffled KFold picks RandomForest_deep off an inflated CV estimate(0.5512) that deploys at 0.4963 on holdout below random.Every time respectinig  method picks LogisticRegression instead , deploying consistently at 0.5189.

**Comment by Shr10122** [comment]
Pushed a commit addressing all the review points:
- Scaler leak fixed via Pipeline, fit per fold
- Sorted by Date before CV (fixes both the shuffle=False baseline 
  and the PurgedGroupKFold monotonic requirement)
- Replaced calendar-based evaluation_times with per-ticker .shift(-5)
- Fixed WalkForwardSplit test_size to reserve an initial training period
- Added the same blocks no purge baseline you suggested
- Rebuilt this as an actual model-selection comparison across 3 candidates

Headline result: shuffled KFold picks RandomForest_deep off an inflated 
CV estimate (0.5512) that deploys at 0.4963 (below random) on the 2024 
holdout. Every time-respecting method  including plain unshuffled 
KFold picks LogisticRegression instead, which deploys at 0.5189 
consistently across all three.
